libraries/Microsoft.Bot.Builder.AI.LanguageGeneration/Analyzer.cs

```diff
@@ -0,0 +1,289 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Text;
+using System.Text.RegularExpressions;
+using Antlr4.Runtime.Misc;
+using Antlr4.Runtime.Tree;
+using Microsoft.Expressions;
+
+namespace Microsoft.Bot.Builder.AI.LanguageGeneration
+{
+
+    /// <summary>
+    /// List all analyzers like dependencies
+    /// 
+    /// </summary>
+    class Analyzer : LGFileParserBaseVisitor<List<string>>
+    {
+        public readonly EvaluationContext Context;
+
+        private Stack<EvaluationTarget> evalutationTargetStack = new Stack<EvaluationTarget>();
+        private EvaluationTarget currentTarget()
+        {
+            // just don't want to write evaluationTargetStack.Peek() everywhere
+            return evalutationTargetStack.Peek();
+        }
+
+
+        public Analyzer(EvaluationContext context)
+        {
+            Context = context;
+        }
+
+        public List<string> AnalyzeTemplate(string templateName)
+        {
+            if (!Context.TemplateContexts.ContainsKey(templateName))
+            {
+                throw new Exception($"No such template: {templateName}");
+            }
+
+            if (evalutationTargetStack.Any(e => e.TemplateName == templateName))
+            {
+                throw new Exception($"Loop deteced: {String.Join(" => ", evalutationTargetStack.Reverse().Select(e => e.TemplateName))} => {templateName}");
+            }
+
+            // Using a stack to track the evalution trace
+            evalutationTargetStack.Push(new EvaluationTarget(templateName, null));
+            var templateResult = Visit(Context.TemplateContexts[templateName]);
```

Comment:
rename templateResult => rawDependencies 
variablesInnerTemplate => parameters

var dependencies = rawDependencies not in parameters // we need to exclude parameters from raw dependencies